File: scgraph/graph_utils.py
from typing import Literal


class GraphUtils:
    def __input_check__(
        self,
        origin_id: int | set[int],
        destination_id: int,
    ) -> None:
        """
        Function:

        - Check that the inputs passed to the shortest path algorithm are valid
        - Raises an exception if the inputs passed are not valid

        Required Arguments:

        - `origin_id`
            - Type: int | set[int]
            - What: The id(s) of the origin node(s) from the graph dictionary to start the shortest path from
        - `destination_id`
            - Type: int
            - What: The id of the destination node from the graph dictionary to end the shortest path at

        Optional Arguments:

        - None
        """
        origin_ids = {origin_id} if isinstance(origin_id, int) else origin_id
        assert isinstance(
            origin_ids, set
        ), "origin_id must be an integer or a set of integers"

        for oid in origin_ids:
            if not isinstance(oid, int) and oid < len(self.graph) and oid >= 0:
                raise Exception(f"Origin node ({oid}) is not in this graph")
        if (
            not isinstance(destination_id, int)
            and origin_id < len(self.graph)
            and origin_id >= 0
        ):
            raise Exception(
                f"Destination node ({destination_id}) is not in this graph"
            )

    def __reconstruct_path__(
        self, destination_id: int, predecessor: list[int]
    ) -> list[int]:
        """
        Function:

        - Reconstruct the shortest path from the destination node to the origin node
        - Return the reconstructed path in the correct order
        - Given the predecessor list, this function reconstructs the path

        Required Arguments:

        - `destination_id`
            - Type: int
            - What: The id of the destination node from the graph dictionary to end the shortest path at
        - `predecessor`
            - Type: list[int]
            - What: The predecessor list that was used to compute the shortest path
            - This list is used to reconstruct the path from the destination node to the origin node
            - Note: Nodes with no predecessor should be -1

        Optional Arguments:

        - None
        """
        output_path = [destination_id]
        while predecessor[destination_id] != -1:
            destination_id = predecessor[destination_id]
            output_path.append(destination_id)
        output_path.reverse()
        return output_path

    def __cycle_check__(self, predecessor_matrix, node_id):
        """
        Function:

        - Check if a cycle exists in the predecessor matrix starting from the given node_id
        - Returns None if no cycle is detected
        - Raises an Exception if a cycle is detected

        Required Arguments:

        - `predecessor_matrix`:
            - Type: list[int]
            - What: A list where the index represents the node id and the value at that index is the predecessor node id
        - `node_id`:
            - Type: int
            - What: The node id to start checking for cycles from

        Optional Arguments:

        - None
        """
        cycle_id = node_id
        while True:
            cycle_id = predecessor_matrix[cycle_id]
            if cycle_id == -1:
                return
            if cycle_id == node_id:
                raise Exception(
                    f"Cycle detected in the graph at node {node_id}"
                )

    def __ensure_inverse_graph__(self) -> list[dict[int, int | float]]:
        """
        Function:

        - Ensure the inverse of the graph as self.inverse_graph is computed and stored
            - The inverse of the graph is a graph where all edges are reversed
            - This is useful for checking the connectivity of the graph and for algorithms that require the inverse graph
        """
        if not hasattr(self, "inverse_graph"):
            self.inverse_graph = [dict() for _ in range(len(self.graph))]
            for origin_id, origin_dict in enumerate(self.graph):
                for destination_id, distance in origin_dict.items():
                    self.inverse_graph[destination_id][origin_id] = distance

    def __connected_check__(self, origin_id: int = 0) -> bool:
        """
        Function:

        - Return True if this graph is fully connected and False if it is not

        Optional Arguments:

        - `origin_id`
            - Type: int
            - What: The id of the origin node from which to start the connectivity check
            - Default: 0
        """
        self.__ensure_inverse_graph__()

        visited = [0] * len(self.graph)
        open_leaves = [origin_id]

        while open_leaves:
            current_id = open_leaves.pop()
            visited[current_id] = 1
            for connected_id in self.graph[current_id]:
                if visited[connected_id] == 0:
                    open_leaves.append(connected_id)

        inverse_visited = [0] * len(self.inverse_graph)
        inverse_open_leaves = [origin_id]
        while inverse_open_leaves:
            current_id = inverse_open_leaves.pop()
            inverse_visited[current_id] = 1
            for connected_id in self.inverse_graph[current_id]:
                if inverse_visited[connected_id] == 0:
                    inverse_open_leaves.append(connected_id)

        return min(visited) == 1 and min(inverse_visited) == 1

    def __symmetric_check__(self) -> bool:
        """
        Function:

        - Return True if this graph is symmetric and False if it is not
            - A graph is symmetric if for every edge from node A to node B, there is an edge from node B to node A with the same distance
        """
        for origin_id, origin_dict in enumerate(self.graph):
            for destination_id, distance in origin_dict.items():
                if distance != self.graph[destination_id].get(origin_id, None):
                    return False
        return True

    def validate(
        self,
        check_symmetry: bool = True,
        check_connected: bool = True,
    ) -> None:
        """
        Function:

        - Validate that this graph object is properly formatted

        Optional Arguments:

        - `check_symmetry`
            - Type: bool
            - What: Whether to check that the graph is symmetric
            - Default: True
        - `check_connected`
            - Type: bool
            - What: Whether to check that the graph is fully connected
            - Default: True
        """
        assert isinstance(self.graph, list), "Your graph must be a list"
        len_graph = len(self.graph)
        if len_graph == 0:
            raise Exception("The provided graph must contain at least one node")
        for origin_id, origin_dict in enumerate(self.graph):
            assert isinstance(
                origin_dict, dict
            ), f"Your graph must be a list of dictionaries but the value for origin {origin_id} is not a dictionary"
            destinations = list(origin_dict.keys())
            lengths = list(origin_dict.values())
            assert all(
                [
                    (isinstance(i, int) and i >= 0 and i < len_graph)
                    for i in destinations
                ]
            ), f"Destination ids must be non-negative integers and equivalent to an existing index, but graph[{origin_id}] has an error in the destination ids"
            assert all(
                [(isinstance(i, (int, float)) and i >= 0) for i in lengths]
            ), f"Distances must be integers or floats, but graph[{origin_id}] contains a non-integer or non-float distance"

        if check_symmetry:
            assert (
                self.__symmetric_check__()
            ), "The provided graph is not symmetric"
        if check_connected:
            assert (
                self.__connected_check__()
            ), "The provided graph is not fully connected"

    def reset_cache(self) -> None:
        """
        Function:

        - Reset the cached shortest path trees
            - This is useful if the graph has been modified and the cached shortest path trees are no longer valid
        """
        self.__cache__ = [0] * len(self.graph)

    def get_cache(self) -> list[dict | None]:
        """
        Function:

        - Get the cached shortest path trees
            - This is useful for saving the cache to disk or for inspecting the cache

        Returns:

        - A list of cached shortest path trees, where each tree is represented as a dictionary
            - If a tree is not cached for a specific node, the corresponding entry in the list will be 0
        """
        return self.__cache__

    def set_cache(self, new_cache: list[dict | Literal[0]]) -> None:
        """
        Function:

        - Set the cached shortest path trees
            - This is useful for loading a cache from disk or for manually setting the cache

        Required Arguments:

        - `new_cache`:
            - Type: list[dict | None]
            - What: A list of cached shortest path trees, where each tree is represented as a dictionary
                - If a tree is not cached for a specific node, the corresponding entry in the list should be 0
        """
        assert isinstance(new_cache, list), "Cache must be a list"
        assert len(new_cache) == len(
            self.graph
        ), "Cache must be the same length as the graph"
        self.__cache__ = new_cache

    def __get__(self, idx: int) -> dict[int, int | float]:
        """
        Function:

        - Get the adjacency dictionary for a specific node in the graph

        Required Arguments:

        - `idx`:
            - Type: int
            - What: The id of the node to get the adjacency dictionary for

        Returns:

        - The adjacency dictionary for the specified node
        """
        return self.graph[idx]


class GraphModifiers:
    def add_node(
        self, node_dict: dict[int, int | float] = None, symmetric: bool = False
    ) -> int:
        """
        Function:

        - Add a node to the graph

        Optional Arguments:

        - `node_dict`:
            - Type: dict[int, int | float]
            - What: Arcs leaving this node.
                - A dictionary where the keys are destination node ids and the values are the distances to those nodes
            - Default: An empty dictionary
        - `symmetric`:
            - Type: bool
            - What: Whether to add edges symmetrically (i.e., add edges from the new node to existing nodes and from existing nodes to the new node)
            - Default: False

        Returns:

        - The id of the newly added node
        """
        node_dict = node_dict if node_dict is not None else dict()
        self.graph.append(node_dict)
        new_node_id = len(self.graph) - 1
        if symmetric:
            for dest_id, distance in node_dict.items():
                self.graph[dest_id][new_node_id] = distance
        return new_node_id

    def add_edge(
        self,
        origin_id: int,
        destination_id: int,
        distance: int | float,
        symmetric: bool = False,
    ) -> None:
        """
        Function:

        - Add an edge to the graph

        Required Arguments:

        - `origin_id`:
            - Type: int
            - What: The id of the origin node
        - `destination_id`:
            - Type: int
            - What: The id of the destination node
        - `distance`:
            - Type: int | float
            - What: The distance between the origin and destination nodes
        - `symmetric`:
            - Type: bool
            - What: Whether to add the edge symmetrically (i.e., add an edge from destination to origin as well)
            - Default: False
        """
        assert origin_id < len(self.graph), "Origin node id is not in the graph"
        assert destination_id < len(
            self.graph
        ), "Destination node id is not in the graph"
        self.graph[origin_id][destination_id] = distance
        if symmetric:
            self.graph[destination_id][origin_id] = distance

    def remove_node(
        self, symmetric_node: bool = False
    ) -> dict[int, int | float]:
        """
        Function:

        - Removes the last node from the graph

        Optional Arguments:

        - `symmetric_node`:
            - Type: bool
            - What: Whether the node being removed has symmetric edges
                - Specifically: This should be True only if all inbound edges to this node are also outbound edges from this node
            - Default: False
            - If True, only uses outbound edges from this node to identify inbound edges to remove from other nodes.

        Returns:

        - The dictionary of edges for this node that were removed from the graph
        """
        assert len(self.graph) > 0, "Graph is empty, cannot remove node"
        node_id = len(self.graph) - 1
        if symmetric_node:
            for dest_id in self.graph[node_id].keys():
                self.graph[dest_id].pop(node_id, None)
        else:
            for origin_dict in self.graph:
                if node_id in origin_dict:
                    origin_dict.pop(node_id)
        return self.graph.pop()

    def remove_edge(
        self, origin_id: int, destination_id: int, symmetric: bool = False
    ) -> int | float | None:
        """
        Function:

        - Remove an edge from the graph

        Required Arguments:

        - `origin_id`:
            - Type: int
            - What: The id of the origin node
        - `destination_id`:
            - Type: int
            - What: The id of the destination node
        - `symmetric`:
            - Type: bool
            - What: Whether to remove the edge symmetrically (i.e., remove the edge from destination to origin as well)
            - Default: False

        Returns:

        - The distance of the removed edge from origin to destination, or None if the edge did not exist
        """
        assert origin_id < len(self.graph), "Origin node id is not in the graph"
        assert destination_id < len(
            self.graph
        ), "Destination node id is not in the graph"
        if symmetric:
            self.graph[destination_id].pop(origin_id, None)
        return self.graph[origin_id].pop(destination_id, None)

File: scgraph/cpp/src/graph_utils.cpp
#include "graph_utils.hpp"
#include <stdexcept>
#include <algorithm>
#include <string>

std::set<int> get_origin_ids(const std::variant<int, std::set<int>>& origin_id) {
    if (std::holds_alternative<int>(origin_id)) {
        return {std::get<int>(origin_id)};
    }
    return std::get<std::set<int>>(origin_id);
}

std::vector<std::vector<std::pair<int, double>>> GraphUtils::serialize_graph(
    const std::vector<std::unordered_map<int, double>>& input_graph) {
    std::vector<std::vector<std::pair<int, double>>> serialized;
    serialized.reserve(input_graph.size());

    for (const auto& node_dict : input_graph) {
        std::vector<std::pair<int, double>> edges;
        edges.reserve(node_dict.size());
        for (const auto& [dest, dist] : node_dict) {
            edges.emplace_back(dest, dist);
        }
        serialized.push_back(std::move(edges));
    }

    return serialized;
}

std::unordered_map<int, double> GraphUtils::get_adjacency_dict(int idx) const {
    if (idx < 0 || idx >= static_cast<int>(graph.size())) {
        throw std::out_of_range("Index out of range");
    }

    std::unordered_map<int, double> result;
    for (const auto& [dest, dist] : graph[idx]) {
        result[dest] = dist;
    }
    return result;
}

void GraphUtils::input_check(const std::variant<int, std::set<int>>& origin_id, int destination_id) const {
    auto origin_ids = get_origin_ids(origin_id);

    for (int oid : origin_ids) {
        if (oid < 0 || oid >= static_cast<int>(graph.size())) {
            throw std::invalid_argument("Origin node (" + std::to_string(oid) + ") is not in this graph");
        }
    }

    if (destination_id < 0 || destination_id >= static_cast<int>(graph.size())) {
        throw std::invalid_argument("Destination node (" + std::to_string(destination_id) + ") is not in this graph");
    }
}

std::vector<int> GraphUtils::reconstruct_path(int destination_id, const std::vector<int>& predecessor) const {
    std::vector<int> output_path;
    output_path.push_back(destination_id);

    while (predecessor[destination_id] != -1) {
        destination_id = predecessor[destination_id];
        output_path.push_back(destination_id);
    }

    std::reverse(output_path.begin(), output_path.end());
    return output_path;
}

void GraphUtils::cycle_check(const std::vector<int>& predecessor_matrix, int node_id) const {
    int cycle_id = node_id;
    while (true) {
        cycle_id = predecessor_matrix[cycle_id];
        if (cycle_id == -1) {
            return;
        }
        if (cycle_id == node_id) {
            throw std::runtime_error("Cycle detected in the graph at node " + std::to_string(node_id));
        }
    }
}

void GraphUtils::ensure_inverse_graph() {
    if (inverse_graph_computed) return;
    inverse_graph.assign(graph.size(), {});
    for (size_t origin_id = 0; origin_id < graph.size(); ++origin_id) {
        for (const auto& [dest_id, distance] : graph[origin_id]) {
            inverse_graph[dest_id].emplace_back(static_cast<int>(origin_id), distance);
        }
    }
    inverse_graph_computed = true;
}

bool GraphUtils::connected_check(int origin_id) {
    ensure_inverse_graph();

    // Forward traversal
    std::vector<int> visited(graph.size(), 0);
    std::vector<int> open_leaves = {origin_id};
    while (!open_leaves.empty()) {
        int current_id = open_leaves.back();
        open_leaves.pop_back();
        visited[current_id] = 1;
        for (const auto& [connected_id, _] : graph[current_id]) {
            if (visited[connected_id] == 0) {
                open_leaves.push_back(connected_id);
            }
        }
    }

    // Inverse traversal
    std::vector<int> inverse_visited(inverse_graph.size(), 0);
    std::vector<int> inverse_open_leaves = {origin_id};
    while (!inverse_open_leaves.empty()) {
        int current_id = inverse_open_leaves.back();
        inverse_open_leaves.pop_back();
        inverse_visited[current_id] = 1;
        for (const auto& [connected_id, _] : inverse_graph[current_id]) {
            if (inverse_visited[connected_id] == 0) {
                inverse_open_leaves.push_back(connected_id);
            }
        }
    }

    return *std::min_element(visited.begin(), visited.end()) == 1 &&
           *std::min_element(inverse_visited.begin(), inverse_visited.end()) == 1;
}

bool GraphUtils::symmetric_check() const {
    for (size_t origin_id = 0; origin_id < graph.size(); ++origin_id) {
        for (const auto& [dest_id, distance] : graph[origin_id]) {
            bool found = false;
            for (const auto& [rev_dest, rev_dist] : graph[dest_id]) {
                if (rev_dest == static_cast<int>(origin_id)) {
                    if (rev_dist != distance) return false;
                    found = true;
                    break;
                }
            }
            if (!found) return false;
        }
    }
    return true;
}

void GraphUtils::validate(bool check_symmetry, bool check_connected) {
    size_t len_graph = graph.size();
    if (len_graph == 0) {
        throw std::invalid_argument("The provided graph must contain at least one node");
    }
    for (size_t origin_id = 0; origin_id < len_graph; ++origin_id) {
        for (const auto& [dest_id, distance] : graph[origin_id]) {
            if (dest_id < 0 || dest_id >= static_cast<int>(len_graph)) {
                throw std::invalid_argument("Destination id " + std::to_string(dest_id) +
                    " at origin " + std::to_string(origin_id) + " is invalid");
            }
            if (distance < 0) {
                throw std::invalid_argument("Distance must be non-negative at origin " +
                    std::to_string(origin_id));
            }
        }
    }

    if (check_symmetry && !symmetric_check()) {
        throw std::invalid_argument("The provided graph is not symmetric");
    }
    if (check_connected && !connected_check()) {
        throw std::runtime_error("The provided graph is not fully connected");
    }
}

void GraphUtils::reset_cache() {
    cache.clear();
    cache.resize(graph.size());
}

const std::unordered_map<int, double> GraphUtils::get(int idx) const {
    return get_adjacency_dict(idx);
}

double GraphUtils::get_path_weight(const std::vector<int>& path) const {
    double total = 0.0;
    for (size_t i = 0; i + 1 < path.size(); ++i) {
        int origin = path[i];
        int dest = path[i + 1];
        bool found = false;
        for (const auto& [d, w] : graph[origin]) {
            if (d == dest) {
                total += w;
                found = true;
                break;
            }
        }
        if (!found) {
            throw std::invalid_argument(
                "No edge from node " + std::to_string(origin) +
                " to node " + std::to_string(dest)
            );
        }
    }
    return total;
}

const std::vector<std::unordered_map<int, double>> GraphUtils::get_graph() const {
    std::vector<std::unordered_map<int, double>> result;
    result.reserve(graph.size());

    for (size_t i = 0; i < graph.size(); ++i) {
        result.push_back(get_adjacency_dict(i));
    }

    return result;
}

int GraphUtils::add_node(const std::unordered_map<int, double>& node_dict, bool symmetric) {
    std::vector<std::pair<int, double>> edges;
    edges.reserve(node_dict.size());
    for (const auto& [dest_id, distance] : node_dict) {
        edges.emplace_back(dest_id, distance);
    }

    graph.push_back(std::move(edges));
    int new_node_id = graph.size() - 1;

    if (symmetric) {
        for (const auto& [dest_id, distance] : node_dict) {
            graph[dest_id].emplace_back(new_node_id, distance);
        }
    }

    reset_cache();
    return new_node_id;
}

void GraphUtils::add_edge(int origin_id, int destination_id, double distance, bool symmetric) {
    if (origin_id >= static_cast<int>(graph.size())) {
        throw std::out_of_range("Origin node id is not in the graph");
    }
    if (destination_id >= static_cast<int>(graph.size())) {
        throw std::out_of_range("Destination node id is not in the graph");
    }

    bool found = false;
    for (auto& [dest, dist] : graph[origin_id]) {
        if (dest == destination_id) {
            dist = distance;
            found = true;
            break;
        }
    }
    if (!found) {
        graph[origin_id].emplace_back(destination_id, distance);
    }

    if (symmetric) {
        found = false;
        for (auto& [dest, dist] : graph[destination_id]) {
            if (dest == origin_id) {
                dist = distance;
                found = true;
                break;
            }
        }
        if (!found) {
            graph[destination_id].emplace_back(origin_id, distance);
        }
    }

    reset_cache();
}

std::unordered_map<int, double> GraphUtils::remove_node(bool symmetric_node) {
    if (graph.empty()) {
        throw std::runtime_error("Graph is empty, cannot remove node");
    }

    int node_id = graph.size() - 1;

    if (symmetric_node) {
        for (const auto& [dest_id, _] : graph[node_id]) {
            auto& dest_edges = graph[dest_id];
            dest_edges.erase(
                std::remove_if(dest_edges.begin(), dest_edges.end(),
                    [node_id](const std::pair<int, double>& edge) {
                        return edge.first == node_id;
                    }),
                dest_edges.end()
            );
        }
    } else {
        for (auto& node_edges : graph) {
            node_edges.erase(
                std::remove_if(node_edges.begin(), node_edges.end(),
                    [node_id](const std::pair<int, double>& edge) {
                        return edge.first == node_id;
                    }),
                node_edges.end()
            );
        }
    }

    std::unordered_map<int, double> removed_dict;
    for (const auto& [dest, dist] : graph[node_id]) {
        removed_dict[dest] = dist;
    }

    graph.pop_back();
    reset_cache();

    return removed_dict;
}

std::optional<double> GraphUtils::remove_edge(int origin_id, int destination_id, bool symmetric) {
    if (origin_id >= static_cast<int>(graph.size())) {
        throw std::out_of_range("Origin node id is not in the graph");
    }
    if (destination_id >= static_cast<int>(graph.size())) {
        throw std::out_of_range("Destination node id is not in the graph");
    }

    std::optional<double> result;

    auto& origin_edges = graph[origin_id];
    for (auto it = origin_edges.begin(); it != origin_edges.end(); ++it) {
        if (it->first == destination_id) {
            result = it->second;
            origin_edges.erase(it);
            break;
        }
    }

    if (symmetric) {
        auto& dest_edges = graph[destination_id];
        dest_edges.erase(
            std::remove_if(dest_edges.begin(), dest_edges.end(),
                [origin_id](const std::pair<int, double>& edge) {
                    return edge.first == origin_id;
                }),
            dest_edges.end()
        );
    }

    reset_cache();
    return result;
}
